dropshot/examples/multiple-servers.rs

```diff
@@ -0,0 +1,377 @@
+ */
+
+#[derive(Debug, Serialize, JsonSchema)]
+struct ServerDescription {
+    name: String,
+    bind_addr: SocketAddr,
+}
+
+/**
+ * Fetch the current list of running servers.
+ */
+#[endpoint {
+    method = GET,
+    path = "/servers",
+}]
+async fn api_get_servers(
+    rqctx: Arc<RequestContext<MultiServerContext>>,
+) -> Result<HttpResponseOk<Vec<ServerDescription>>, HttpError> {
+    let api_context = rqctx.context();
+
+    let servers = api_context.shared.servers.lock().await;
+    let servers = servers
+        .iter()
+        .map(|(name, server)| ServerDescription {
+            name: name.clone(),
+            bind_addr: server.local_addr(),
+        })
+        .collect();
+
+    Ok(HttpResponseOk(servers))
+}
+
+#[derive(Deserialize, JsonSchema)]
+struct PathName {
+    name: String,
+}
+
+/**
+ * Start a new running server.
+ */
+#[endpoint {
+    method = POST,
+    path = "/servers/{name}",
+}]
+async fn api_start_server(
+    rqctx: Arc<RequestContext<MultiServerContext>>,
+    path: Path<PathName>,
+    body: TypedBody<SocketAddr>,
+) -> Result<HttpResponseCreated<ServerDescription>, HttpError> {
+    let api_context = rqctx.context();
+    let name = path.into_inner().name;
+    let bind_addr = body.into_inner();
+
+    api_context.shared.start_server(&name, bind_addr).await.map_err(|err| {
+        /*
+         * `for_bad_request` _might_ not be right (e.g., we might have some
+         * spurious OS error starting the server), but it's likely right (the
+         * most likely cause for failure is a duplicate name or already-in-use
+         * bind address), and we're being lazy with errors in this example.
+         */
+        HttpError::for_bad_request(
+            Some("StartServerFailed".to_string()),
+            format!("failed to start server {name:?}: {err}"),
+        )
+    })?;
+
+    Ok(HttpResponseCreated(ServerDescription { name, bind_addr }))
+}
+
+/**
+ * Stop a running server by name.
+ */
+#[endpoint {
+    method = DELETE,
+    path = "/servers/{name}",
+}]
+async fn api_stop_server(
+    rqctx: Arc<RequestContext<MultiServerContext>>,
+    path: Path<PathName>,
+) -> Result<HttpResponseDeleted, HttpError> {
+    let api_context = rqctx.context();
+    let name = path.into_inner().name;
+
+    let mut servers = api_context.shared.servers.lock().await;
+    let server = servers.remove(&name).ok_or_else(|| {
+        HttpError::for_bad_request(
+            Some("InvalidServerName".to_string()),
+            format!("no server named {name:?}"),
+        )
+    })?;
+
+    /*
+     * We want to shut down `server`, but it might be the very server handling
+     * this request! Move the shutdown onto a background task to allow this
+     * request to complete; otherwise, we deadlock.
+     *
+     * We can safely discard the result of `close()` because `main()` also gets
+     * it via the shutdown handle it received when `server` was created.
+     */
+    tokio::spawn(server.close());
```

Comment:
After https://github.com/oxidecomputer/dropshot/pull/537 , I wonder if we should no longer make `close` wait for the server to terminate.

We could add a convenience `close_and_wait` for anyone that wants backwards compatibility?

Reply:
Changing the behavior of `close()` seems a little dicey from a backwards compatibility standpoint, but I do see your point. A couple other ideas, although I'm not sure I like these all that much more:

1. Deprecate `close()` and add _two_ new methods: `close_and_wait()` + `close_nowait()`.
2. Change `close()` to take an argument specifying whether or not to wait: `close(CloseBehavior::Wait)` or somesuch. This would be backwards incompatible in a way that old clients would have to update instead of silently getting different behavior.